prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -352,7 +361,95 @@ instance Pretty Lazy.Text where pretty = pretty . Lazy.toStrict
 -- []
 instance Pretty Void where pretty = absurd
 
+-- | Overloaded conversion to 'Doc', lifted to unary type constructors.
+--
+-- This is most useful for:
+-- 1. defining 'Pretty' instances for recursive types,
+-- 2. defining 'Pretty' instances for type constructors without 'Functor'
+--    instances, and
+-- 3. efficiently pretty-printing type constructors with 'Functor' instances
+--    whose 'fmap' traverses the whole structure.
+--
+-- Laws:
+--
+--   1. output should be pretty. :-)
+class Pretty1 f where
+
+    -- | >>> liftPretty (parens . pretty) (list . map (parens . pretty)) (Just "hello")
+    -- (hello)
+    liftPretty
+        :: (a -> Doc ann)   -- ^ A function to print a single value.
+        -> ([a] -> Doc ann) -- ^ A function to print a list. Used for [].
```

Comment:
Is this really necessary? It clutters the definitions quite a bit, but it’s usually just `list . map pretty`.

Reply:
It’s necessary for the `[]` definition to work as expected for `Char`, unfortunately.

Reply:
Arrr, I see.

Actually, this brings up a new law: `Pretty1 f` and `Pretty (f a)` should result in identical behavior! :-)